# Introduction

 As part of my Data Analytics course, I am currently working on a project using a retail dataset sourced from Kaggle.

The dataset comprises three interrelated CSV files, a Features data set.csv, a Stores data-set.csv, and a Sales data-set.csv.
The Features data set which I will start with below contains 12 columns detailing regional and contextual information for each store, including metrics such as temperature, fuel prices, and promotional activity, recorded across various dates.

In [1]:
import pandas as pd

In [4]:
df = pd.read_csv('../data/inputs/raw/Features data set.csv')
df

Store        Date  Temperature  Fuel_Price  MarkDown1  MarkDown2  \
0         1  05/02/2010        42.31       2.572        NaN        NaN   
1         1  12/02/2010        38.51       2.548        NaN        NaN   
2         1  19/02/2010        39.93       2.514        NaN        NaN   
3         1  26/02/2010        46.63       2.561        NaN        NaN   
4         1  05/03/2010        46.50       2.625        NaN        NaN   
...     ...         ...          ...         ...        ...        ...   
8185     45  28/06/2013        76.05       3.639    4842.29     975.03   
8186     45  05/07/2013        77.50       3.614    9090.48    2268.58   
8187     45  12/07/2013        79.37       3.614    3789.94    1827.31   
8188     45  19/07/2013        82.84       3.737    2961.49    1047.07   
8189     45  26/07/2013        76.06       3.804     212.02     851.73   

      MarkDown3  MarkDown4  MarkDown5         CPI  Unemployment  IsHoliday  
0           NaN        NaN        NaN  211.096358         8.106      False  
1           NaN        NaN        NaN  211.242170         8.106       True  
2           NaN        NaN        NaN  211.289143         8.106      False  
3           NaN        NaN        NaN  211.319643         8.106      False  
4           NaN        NaN        NaN  211.350143         8.106      False  
...         ...        ...        ...         ...           ...        ...  
8185       3.00    2449.97    3169.69         NaN           NaN      False  
8186     582.74    5797.47    1514.93         NaN           NaN      False  
8187      85.72     744.84    2150.36         NaN           NaN      False  
8188     204.19     363.00    1059.46         NaN           NaN      False  
8189       2.06      10.88    1864.57         NaN           NaN      False  

[8190 rows x 12 columns]

There are missing values or NaN in all of the MarkDown columns which I will look at handling by using zero for MarkDown columns which will ensure a stable trend analysis while preserving the logical meaning.  - The use of Median in CPI will provides a more stable central value, keeping our analysis accurate.



In [14]:
df = pd.read_csv('../data/inputs/raw/Features data set.csv')
df.fillna({'CPI': df['CPI'].median(), 'MarkDown1': 0, 'MarkDown2': 0, 'MarkDown3': 0, 'MarkDown4': 0, 'MarkDown5': 0}, inplace=True)
df.head()  

Store        Date  Temperature  Fuel_Price  MarkDown1  MarkDown2  \
0      1  05/02/2010        42.31       2.572        0.0        0.0   
1      1  12/02/2010        38.51       2.548        0.0        0.0   
2      1  19/02/2010        39.93       2.514        0.0        0.0   
3      1  26/02/2010        46.63       2.561        0.0        0.0   
4      1  05/03/2010        46.50       2.625        0.0        0.0   

   MarkDown3  MarkDown4  MarkDown5         CPI  Unemployment  IsHoliday  
0        0.0        0.0        0.0  211.096358         8.106      False  
1        0.0        0.0        0.0  211.242170         8.106       True  
2        0.0        0.0        0.0  211.289143         8.106      False  
3        0.0        0.0        0.0  211.319643         8.106      False  
4        0.0        0.0        0.0  211.350143         8.106      False

Next we will look at standardizing the date format ensures consistency which will allow for easier manipulation when analyzing time-based trends in the retail dataset. Also by converting the date format it will ensure correct recognition by Pandas.



In [17]:
df['Date'] = pd.to_datetime(df['Date'], format='%d/%m/%Y')
df.head()

Store       Date  Temperature  Fuel_Price  MarkDown1  MarkDown2  MarkDown3  \
0      1 2010-02-05        42.31       2.572        0.0        0.0        0.0   
1      1 2010-02-12        38.51       2.548        0.0        0.0        0.0   
2      1 2010-02-19        39.93       2.514        0.0        0.0        0.0   
3      1 2010-02-26        46.63       2.561        0.0        0.0        0.0   
4      1 2010-03-05        46.50       2.625        0.0        0.0        0.0   

   MarkDown4  MarkDown5         CPI  Unemployment  IsHoliday  
0        0.0        0.0  211.096358         8.106      False  
1        0.0        0.0  211.242170         8.106       True  
2        0.0        0.0  211.289143         8.106      False  
3        0.0        0.0  211.319643         8.106      False  
4        0.0        0.0  211.350143         8.106      False

Now we will look at converting the numeric columns such as Fuel_Price into a float ensuring accurate analysis and prevent any calculation errors. As it will stop it from being mistakenly stored as a string, where any mathematical operations will fail.

In [18]:
df['Fuel_Price'] = df['Fuel_Price'].astype(float)
df.head()

Store       Date  Temperature  Fuel_Price  MarkDown1  MarkDown2  MarkDown3  \
0      1 2010-02-05        42.31       2.572        0.0        0.0        0.0   
1      1 2010-02-12        38.51       2.548        0.0        0.0        0.0   
2      1 2010-02-19        39.93       2.514        0.0        0.0        0.0   
3      1 2010-02-26        46.63       2.561        0.0        0.0        0.0   
4      1 2010-03-05        46.50       2.625        0.0        0.0        0.0   

   MarkDown4  MarkDown5         CPI  Unemployment  IsHoliday  
0        0.0        0.0  211.096358         8.106      False  
1        0.0        0.0  211.242170         8.106       True  
2        0.0        0.0  211.289143         8.106      False  
3        0.0        0.0  211.319643         8.106      False  
4        0.0        0.0  211.350143         8.106      False

Lets take a look at an overview of all columns, including their data types and missing values and also to make sure that the fuel price has converted to a float.

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8190 entries, 0 to 8189
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Store         8190 non-null   int64         
 1   Date          8190 non-null   datetime64[ns]
 2   Temperature   8190 non-null   float64       
 3   Fuel_Price    8190 non-null   float64       
 4   MarkDown1     8190 non-null   float64       
 5   MarkDown2     8190 non-null   float64       
 6   MarkDown3     8190 non-null   float64       
 7   MarkDown4     8190 non-null   float64       
 8   MarkDown5     8190 non-null   float64       
 9   CPI           8190 non-null   float64       
 10  Unemployment  7605 non-null   float64       
 11  IsHoliday     8190 non-null   bool          
dtypes: bool(1), datetime64[ns](1), float64(9), int64(1)
memory usage: 712.0 KB


From the dataset above we can see that the Unemployment column has missing values (7605 non-null out of 8190), meaning nearly 7% of the rows are incomplete. We will look at using Median to ensure that the data stays stable and following the same logic we used in CPI above.


In [22]:
df['Unemployment'].fillna(df['Unemployment'].median(), inplace=True)
df.info()
df.head()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8190 entries, 0 to 8189
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Store         8190 non-null   int64         
 1   Date          8190 non-null   datetime64[ns]
 2   Temperature   8190 non-null   float64       
 3   Fuel_Price    8190 non-null   float64       
 4   MarkDown1     8190 non-null   float64       
 5   MarkDown2     8190 non-null   float64       
 6   MarkDown3     8190 non-null   float64       
 7   MarkDown4     8190 non-null   float64       
 8   MarkDown5     8190 non-null   float64       
 9   CPI           8190 non-null   float64       
 10  Unemployment  8190 non-null   float64       
 11  IsHoliday     8190 non-null   bool          
dtypes: bool(1), datetime64[ns](1), float64(9), int64(1)
memory usage: 712.0 KB


Store       Date  Temperature  Fuel_Price  MarkDown1  MarkDown2  MarkDown3  \
0      1 2010-02-05        42.31       2.572        0.0        0.0        0.0   
1      1 2010-02-12        38.51       2.548        0.0        0.0        0.0   
2      1 2010-02-19        39.93       2.514        0.0        0.0        0.0   
3      1 2010-02-26        46.63       2.561        0.0        0.0        0.0   
4      1 2010-03-05        46.50       2.625        0.0        0.0        0.0   

   MarkDown4  MarkDown5         CPI  Unemployment  IsHoliday  
0        0.0        0.0  211.096358         8.106      False  
1        0.0        0.0  211.242170         8.106       True  
2        0.0        0.0  211.289143         8.106      False  
3        0.0        0.0  211.319643         8.106      False  
4        0.0        0.0  211.350143         8.106      False

Now lets look at removing duplicates, it is important to remove duplicates from datasets to maintain accuracy and reliability of data and not cause inflation of certain values like cost or sales and ensures the data is trustworthy.

In [25]:
df.drop_duplicates(inplace=True)

We have now completed Data cleansing of the Features Data set and have saved as cleaned retail data.

In [ ]:
import os

os.makedirs('data/processed', exist_ok=True)

df.to_csv('data/processed/cleaned_retail_data.csv', index=False)

We will now move onto the sales data-set which consists of 5 columns capturing weekly sales figures for each department within each store.

In [33]:
import pandas as pd

In [43]:
sales_df.describe()
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 421570 entries, 0 to 421569
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   Store         421570 non-null  int64         
 1   Dept          421570 non-null  int64         
 2   Date          421570 non-null  datetime64[ns]
 3   Weekly_Sales  421570 non-null  float64       
 4   IsHoliday     421570 non-null  bool          
dtypes: bool(1), datetime64[ns](1), float64(1), int64(2)
memory usage: 13.3 MB


In [38]:
sales_df = pd.read_csv('../data/inputs/raw/sales data-set.csv')
print(sales_df.head())


   Store  Dept        Date  Weekly_Sales  IsHoliday
0      1     1  05/02/2010      24924.50      False
1      1     1  12/02/2010      46039.49       True
2      1     1  19/02/2010      41595.55      False
3      1     1  26/02/2010      19403.54      False
4      1     1  05/03/2010      21827.90      False


As before in Features data set I will begin by checking for missing values.

In [39]:
sales_df.isnull().sum()

Store           0
Dept            0
Date            0
Weekly_Sales    0
IsHoliday       0
dtype: int64

The results show unlike the Features data set the sales data-set has has zero missing values, we can now move on to stadardizing the Date time format to ensure easier manipulation of data.

Next we will converts Date into a proper datetime format for easy time-based analysis


In [42]:
sales_df['Date'] = pd.to_datetime(sales_df['Date'], format='%d/%m/%Y')
sales_df.head()

Store  Dept       Date  Weekly_Sales  IsHoliday
0      1     1 2010-02-05      24924.50      False
1      1     1 2010-02-12      46039.49       True
2      1     1 2010-02-19      41595.55      False
3      1     1 2010-02-26      19403.54      False
4      1     1 2010-03-05      21827.90      False

To make sure that our weekly sales can be used in calculations,and stored correctly as a float for data integrity and analysis accuracy.

In [45]:
sales_df['Weekly_Sales'] = sales_df['Weekly_Sales'].astype(float)
sales_df.head()
sales_df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 421570 entries, 0 to 421569
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   Store         421570 non-null  int64         
 1   Dept          421570 non-null  int64         
 2   Date          421570 non-null  datetime64[ns]
 3   Weekly_Sales  421570 non-null  float64       
 4   IsHoliday     421570 non-null  bool          
dtypes: bool(1), datetime64[ns](1), float64(1), int64(2)
memory usage: 13.3 MB


Check the data-set to see if Weekly_Sales has missing values,and then fill them with median like before.



In [47]:
sales_df['Weekly_Sales'].fillna(sales_df['Weekly_Sales'].median(), inplace=True)
sales_df.head()

Store  Dept       Date  Weekly_Sales  IsHoliday
0      1     1 2010-02-05      24924.50      False
1      1     1 2010-02-12      46039.49       True
2      1     1 2010-02-19      41595.55      False
3      1     1 2010-02-26      19403.54      False
4      1     1 2010-03-05      21827.90      False

We will now look for any identifying and unexpected values that could indicate errors or anomalies in the dataset


In [48]:
print(sales_df['Dept'].unique())  # See all unique department numbers
print(sales_df['Store'].unique())  # See all unique store IDs

[ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 16 17 18 19 20 21 22 23 24 25
 26 27 28 29 30 31 32 33 34 35 36 37 38 40 41 42 44 45 46 47 48 49 51 52
 54 55 56 58 59 60 67 71 72 74 77 78 79 80 81 82 83 85 87 90 91 92 93 94
 95 96 97 98 99 39 50 43 65]
[ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45]


Now we can look at detecting and removing any duplicates.

In [52]:
sales_df.drop_duplicates(inplace=True)
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 243072 entries, 0 to 417657
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   Store         243072 non-null  int64         
 1   Dept          243072 non-null  int64         
 2   Date          243072 non-null  datetime64[ns]
 3   Weekly_Sales  243072 non-null  float64       
 4   IsHoliday     243072 non-null  bool          
dtypes: bool(1), datetime64[ns](1), float64(1), int64(2)
memory usage: 9.5 MB


In [53]:
print(sales_df.duplicated().sum())  # Should return 0 if there are no duplicates

0
